Add documentation about credentials for Organization Folders and Multibranch Pipelines (#7770)
jenkins-infra/jenkins.io: content/doc/book/security/securing-org-folders-and-multibranch-pipelines.adoc

* Depending on your SCM provider's credential-related features, you may not be able to use Organization Folders securely, and would instead need to create one Multibranch Pipeline for each repository with a credential defined on it that only allows access to that specific repository.

Unfortunately, if your primary concern is to prevent users with SCM read/write access to a repository from using the scan credentials to gain elevated permissions, there is no generic solution.
In some cases, you may be able to remove permissions from the credential to make it less powerful.
For example, you may be able to remove webhook-related permissions from the credentials, which will prevent Jenkins from automatically managing webhooks, in favor of having an administrator configure webhooks manually.

=== Enhanced GitHub App Credentials

If you are using GitHub Branch Source Plugin with GitHub App Credentials, the plugin offers various features as of version TODO that can be used to improve security.

> See TODO.

Author: I guess the best way to deal with the dependency in documentation links between the PRs would be to temporarily comment this out, merge this so that the `github-branch-source` PR can reference this page, then come back and uncomment this and link to the updated `github-branch-source` docs.

Author: I commented this out in [c237b6b](https://github.com/jenkins-infra/jenkins.io/pull/7770/commits/c237b6b17a250eb7ffa02366ae635a0c26d7566a).

> === Enhanced GitHub App Credentials
> 
> If you are using GitHub Branch Source Plugin with GitHub App Credentials, the plugin offers various features as of version TODO that can be used to improve security.
> See TODO.

Author: ```suggestion
// TODO: Update and uncomment once https://github.com/jenkinsci/github-branch-source-plugin/pull/822 is released.
////
=== Enhanced GitHub App Credentials

If you are using GitHub Branch Source Plugin with GitHub App Credentials, the plugin offers various features as of version TODO that can be used to improve security.
See TODO.
////
```

These projects interact with external SCM providers in various ways, which generally require credentials with SCM provider API access to be made available to Jenkins ("scan credentials").
The following sections discuss potential issues with visibility and accessibility of those credentials.

== Credentials

For many Branch Source configurations, such as those provided by GitHub Branch Source plugin and Bitbucket Branch Source plugin, the Jenkins user who configures the project only needs to provide a single credential, which will be used in two ways:

* By the controller, to perform plugin operations like Organization Folder scans, Multibranch Pipeline branch indexing, webhook modification, commit status updates, and more. These operations generally interact with SCM provider APIs that may require significant permissions beyond read/write access to the contents of a repository.
* By nodes, to check out the repositories being built by the child Pipeline jobs.

> Credentials defined using Global scope are available to all children of the object where they are defined.

Reviewer: ```suggestion
Credentials defined using the Global scope are available to all children of the object where they are defined.
```
Small addition of "the"

This means that any credentials available to Organization Folders and Multibranch Pipelines will also be available to their child Pipeline jobs.

> Any user who has Job/Create permission in a location where a credential is available must be trusted to use that credential arbitrarily (and in many cases, Job/Configure is also enough to use credentials arbitrarily).

Reviewer: ```suggestion
Any user who has Job/Create permission in a location where a credential is available must be trusted to use that credential arbitrarily (and in many cases, Job/Configure is enough to use credentials arbitrarily).
```
Suggestion to remove "also" unless it should absolutely describe that this would only be in addition to something else.

When using Jenkins Pipeline, that trust is extended to any user who is able to modify `Jenkinsfile`, and have their changes used by Jenkins.
Which users are trusted to modify `Jenkinsfile` depends on the Branch Source configuration for the Pipeline, as they are able to modify `Jenkinsfile`.

> For example, any user who can write to a branch in the repository is considered trusted by GitHub Branch Source plugin.

Reviewer: ```suggestion
For example, any user who can write to a branch in the repository is considered trusted by the GitHub Branch Source plugin.
```
Small suggestion to add "the".

====

This may be undesirable in two ways:

* Write access to one repository can be used to obtain credentials that have access to other repositories.
* Write access to one repository can be used to obtain credentials with permissions beyond repository read/write access.

== Affected scenarios

The above concerns may not matter to users with simple security models.

You should consider the above implications carefully if:

* You use a single organization folder in Jenkins to build multiple SCM repositories for which the users with SCM access to one repository must not have access to the other repositories being built.
* Your SCM users must only have read/write access to repositories and no additional permissions.

Note that if you allow untrusted users to create or configure jobs in locations where credentials are available, this issue does not matter, as those users can use the credentials directly.

Additionally, these concerns only apply to users who do not have Job/Configure permission (or more powerful permissions) in Jenkins, but who do have the ability to edit Pipelines, for example because they have write access to repositories that contain Jenkinsfiles.

> If untrusted users do not have the ability to edit Jenkinsfiles, for example because Multibranch Pipelines are configured to use a centralized Pipeline via a plugin such as https://plugins.jenkins.io/inline-pipeline/[inline-pipeline] or https://plugins.jenkins.io/pipeline-multibranch-defaults/[pipeline-multibranch-defaults], these issues do not matter.

Reviewer: ```suggestion
If untrusted users do not have the ability to edit Jenkinsfiles, for example because Multibranch Pipelines are configured to use a centralized Pipeline via a plugin such as plugin:inline-pipeline[Multibranch Pipeline Inline Definition] or plugin:pipeline-multibranch-defaults[Pipeline: Multibranch with defaults], these issues do not matter.
```
Suggestion to utilize the plugin macro and actual names of the plugins.
